### PR TITLE
Ansible runner integration POC

### PR DESCRIPTION
**Pull request by shawn-hurley**
* add ansible runner as the command to run.
* Read the event data to populate a status object
* Create  generic CR Status object.
* update/change the status

### CODE REVIEW

**Comment by ehelms** [comment]
Generally speaking this is working for me when running my setup (10+ roles, 15+ services) with the ansible-runner based operator. I am seeing that the Ansible configuration option `stdout_callback = actionable` does not seem to be respected but that may be an issue with ansible-runner. I am going to try to independently verify that but would not hold this change up on account of it. I do also see status information now.

Is there way to configure custom status? Is that planned?

**Comment by ehelms** [comment]
I dug more into ansible-runner and it uses a custom stdout_callback to emit the event data as files. So I played around a bit how to mimic the behavior I wanted. Here's a diff and sample output to kinda show my thoughts on reducing noise so that the operator gives better feedback when using ansible-runner:

Diff
```
diff --git a/pkg/runner/runner.go b/pkg/runner/runner.go
index db0a2fd..6d5aaa8 100644
--- a/pkg/runner/runner.go
+++ b/pkg/runner/runner.go
@@ -36,6 +36,18 @@ func (e EventTime) MarshalJSON() ([]byte, error) {
        return []byte(fmt.Sprintf("\"%s\"", e.Time.Format("2006-01-02T15:04:05.99999999"))), nil
 }
 
+type StdOut struct {
+       Changed bool    `json:"changed"`
+       Error           int                     `json:"error"`
+       Item            string  `json:"item"`
+       Msg                     string  `json:"msg"`
+}
+
+type EventData struct {
+       Task string     `json:"task"`
+       Role string     `json:"role"`
+}
+
 // JobEvent - event of an ansible run.
 type JobEvent struct {
        UUID      string                 `json:"uuid"`
@@ -44,7 +56,7 @@ type JobEvent struct {
        StartLine int                    `json:"start_line"`
        EndLine   int                    `json:"EndLine"`
        Event     string                 `json:"event"`
-       EventData map[string]interface{} `json:"event_data"`
+       EventData EventData                                              `json:"event_data"`
        PID       int                    `json:"pid"`
        Created   EventTime              `json:"created"`
 }
@@ -106,8 +118,6 @@ func (p *Playbook) Run(parameters map[string]interface{}, name, namespace string
        logrus.Infof("running: %v for playbook: %v", ident, p.Path)
 
        dc := exec.Command("ansible-runner", "-vv", "-p", "playbook.yaml", "-i", fmt.Sprintf("%v", ident), "run", runnerSandbox)
-       dc.Stdout = os.Stdout
-       dc.Stderr = os.Stderr
        err = dc.Run()
        if err != nil {
                return nil, err
@@ -123,6 +133,28 @@ func (p *Playbook) Run(parameters map[string]interface{}, name, namespace string
                return nil, fmt.Errorf("Unable to read event data")
        }
        sort.Sort(fileInfos(eventFiles))
+
+       for i := 0; i < len(eventFiles); i++ {
+               file, _ := ioutil.ReadFile(fmt.Sprintf("%v/artifacts/%v/job_events/%v", runnerSandbox, ident, eventFiles[i].Name()))
+               jobEvent := JobEvent{}
+
+               err = json.Unmarshal(file, &jobEvent)
+
+               if jobEvent.Event == "runner_item_on_changed" || jobEvent.Event == "runner_item_on_failed" {
+                       split := strings.Split(jobEvent.StdOut, "=> ")
+                       stdoutJson := split[1]
+
+                       stdOut := StdOut{}
+                       err = json.Unmarshal([]byte(stdoutJson), &stdOut)
+                       logrus.WithFields(logrus.Fields{
+                               "changed": stdOut.Changed,
+                               "error": stdOut.Error,
+                               "task": jobEvent.EventData.Task,
+                               "role": jobEvent.EventData.Role,
+                       }).Info(stdOut.Msg)
+               }
+       }
+
        //get the last event, which should be a status.
        d, err := ioutil.ReadFile(fmt.Sprintf("%v/artifacts/%v/job_events/%v", runnerSandbox, ident, eventFiles[len(eventFiles)-1].Name()))
        if err != nil {
```

Output:
```
time="2018-07-19T16:44:34Z" level=info msg="Failed to retrieve requested object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"routes \\\"foreman-https\\\" is forbidden: User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in the namespace \\\"foreman\\\": User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in project \\\"foreman\\\"\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"foreman-https\",\"kind\":\"routes\"},\"code\":403}\n" changed=false error=403 role=foreman-routes task="foreman routes"
time="2018-07-19T16:44:34Z" level=info msg="Failed to retrieve requested object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"routes \\\"foreman-http-pub\\\" is forbidden: User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in the namespace \\\"foreman\\\": User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in project \\\"foreman\\\"\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"foreman-http-pub\",\"kind\":\"routes\"},\"code\":403}\n" changed=false error=403 role=foreman-routes task="foreman routes"
time="2018-07-19T16:44:34Z" level=info msg="Failed to retrieve requested object: {\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"routes \\\"foreman-http-pulp\\\" is forbidden: User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in the namespace \\\"foreman\\\": User \\\"system:serviceaccount:foreman:foreman-operator\\\" cannot get routes in project \\\"foreman\\\"\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"foreman-http-pulp\",\"kind\":\"routes\"},\"code\":403}\n" changed=false error=403 role=foreman-routes task="foreman routes"
```

**Comment by ehelms** [comment]
One additional thought with ansible-runner, since it creates data on the system and thus could grow over time as the operator continues to run, have you given thought to rotating job_events off the file system?

**Comment by shawn-hurley** [comment]
@ehelms We have not, but now we will 😄 

I will be adding this to our documentation and feature tracker so we don't lose this :)

**Comment by ehelms** [comment]
@shawn-hurley I think that sounds solid. My only request would be to make it configurable whether to stream all events or only change and failure events to the logger for easier debugging of roles.

File: cmd/ansible-operator/main.go [hunk 1]
```diff
@@ -92,8 +95,11 @@ func runSDK(done chan error) {
 		done <- err
 		return
 	}
+	//Seed rand
```
Comment (fabianvf) [inline cmd/ansible-operator/main.go]: This comment seems a little unnecessary

File: example/Dockerfile [hunk 2]
```diff
@@ -1,7 +1,7 @@
-FROM docker.io/shurley/ansible-operator
+FROM docker.io/shurley/ansible-operator-test
```
Comment (fabianvf) [inline example/Dockerfile]: why `ansible-operator-test`?

File: example/deploy/operator.yaml [hunk 3]
```diff
@@ -14,7 +14,7 @@ spec:
     spec:
```
Comment (fabianvf) [inline example/deploy/operator.yaml]: What's the difference between example/deploy/operator.yaml and deploy/operator.yaml?

File: pkg/kubeconfig/kubeconfig.go [hunk 4]
```diff
@@ -72,12 +72,10 @@ func Create(ownerRef metav1.OwnerReference, proxyURL string, namespace string) (
 	if err != nil {
 		return nil, err
 	}
+	defer file.Close()
```
Comment (fabianvf) [inline pkg/kubeconfig/kubeconfig.go]: Just a note, if this errors do we still want to catch it?

File: pkg/proxy/proxy.go [hunk 5]
```diff
@@ -85,6 +84,7 @@ func injectOwnerReference(h http.Handler) http.Handler {
 			req.Body = ioutil.NopCloser(bytes.NewBuffer(newBody))
 			req.ContentLength = int64(len(newBody))
 		}
+		req.Header.Del("Authorization")
```
Comment (fabianvf) [inline pkg/proxy/proxy.go]: Probably good to add a comment here/document it in general. Might throw some people off.

File: pkg/runner/runner.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+// Run - This should allow the playbook runner to run.
+func (p *Playbook) Run(parameters map[string]interface{}, name, namespace, kc string) (*StatusJobEvent, error) {
+	parameters["meta"] = map[string]string{"namespace": namespace, "name": name}
+	runnerSandbox := fmt.Sprintf("/home/ansible-operator/runner/%s/%s/%s/%s/%s", p.GVK.Group, p.GVK.Version, p.GVK.Kind, namespace, name)
+	b, err := json.Marshal(parameters)
+	if err != nil {
+		return nil, err
+	}
+	//Write parameters to correct file on disk
+	err = createRunnerEnvironment(b, runnerSandbox, kc)
```
Comment (fabianvf) [inline pkg/runner/runner.go]: What is kc? I think some of these short variable names make the code a little hard to follow.

File: pkg/runner/runner.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	//copy playbook from the path to the project of the runnerEnvironment
+	err = copyFile(p.Path, fmt.Sprintf("%v/project/playbook.yaml", runnerSandbox))
+	if err != nil {
+		return nil, err
+	}
+	ident := rand.Int()
+	logrus.Infof("running: %v for playbook: %v", ident, p.Path)
+
+	dc := exec.Command("ansible-runner", "-vv", "-p", "playbook.yaml", "-i", fmt.Sprintf("%v", ident), "run", runnerSandbox)
```
Comment (fabianvf) [inline pkg/runner/runner.go]: Would we want the verbosity to be configurable?

File: pkg/runner/runner.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+	err = createRunnerEnvironment(b, runnerSandbox, kc)
+	if err != nil {
+		return nil, err
+	}
+	//copy playbook from the path to the project of the runnerEnvironment
+	err = copyFile(p.Path, fmt.Sprintf("%v/project/playbook.yaml", runnerSandbox))
+	if err != nil {
+		return nil, err
+	}
+	ident := rand.Int()
```
Comment (fabianvf) [inline pkg/runner/runner.go]: What is `ident` used for?

File: pkg/runner/runner.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		logrus.Errorf("unable to create extravars file - %v", runnerSandbox)
+		return err
+	}
+	return nil
+}
+
+// Copy the src file to dst. Any existing file will be overwritten and will not
+// copy file attributes.
+func copyFile(src, dst string) error {
```
Comment (fabianvf) [inline pkg/runner/runner.go]: Is this not a function provided in the Go standard library?

File: pkg/runner/types.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+type Status struct {
+	Ok               int       `json:"ok"`
+	Changed          int       `json:"changed"`
+	Skipped          int       `json:"skipped"`
+	Failures         int       `json:"failures"`
+	TimeOfCompletion EventTime `json:"completion"`
+}
+
+func NewStatusFromStatusJobEvent(je *StatusJobEvent) Status {
+	o := 0
```
Comment (fabianvf) [inline pkg/runner/types.go]: I think more descriptive variable names would be helpful here.

File: pkg/stub/handler.go [hunk 11]
```diff
@@ -2,25 +2,23 @@ package stub
...
 
-func NewHandler(m map[schema.GroupVersionKind]string) sdk.Handler {
+func NewHandler(m map[schema.GroupVersionKind]runner.Runner) sdk.Handler {
 	return &Handler{crdToPlaybook: m}
 }
 
 type Handler struct {
-	crdToPlaybook map[schema.GroupVersionKind]string
+	crdToPlaybook map[schema.GroupVersionKind]runner.Runner
 	// Fill me
```
Comment (fabianvf) [inline pkg/stub/handler.go]: not a change from this PR, but this is a weird comment

File: pkg/stub/handler.go [hunk 12]
```diff
@@ -56,20 +54,24 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
 		logrus.Warnf("spec is not a map[string]interface{} - %#v", s)
 		return nil
 	}
-
-	return runPlaybook(p, spec, kc.Name())
-}
-
-func runPlaybook(path string, parameters map[string]interface{}, configPath string) error {
-	b, err := json.Marshal(parameters)
+	je, err := p.Run(spec, u.GetName(), u.GetNamespace(), kc.Name())
```
Comment (fabianvf) [inline pkg/stub/handler.go]: What is `je`?

File: pkg/stub/handler.go [hunk 13]
```diff
@@ -56,20 +54,24 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
...
+	if !ok {
+		u.Object["status"] = runner.ResourceStatus{
+			Status: runner.NewStatusFromStatusJobEvent(je),
+		}
+		sdk.Update(u)
+		logrus.Infof("adding status for the first time")
+		return nil
+	}
+	// Need to conver the map[string]interface into a resource status.
+	if update, rs := runner.UpdateResourceStatus(statusMap, je); update {
```
Comment (fabianvf) [inline pkg/stub/handler.go]: `s/rs/status`

File: example/deploy/operator.yaml (hunk 3 above)
Comment (shawn-hurley) [reply example/deploy/operator.yaml]: deploy/operator would only deploy the ansible-operator. example deploy/operator deploys the example operator.

Comment (fabianvf) [reply example/deploy/operator.yaml]: Interesting, the definitions look pretty much the same except for a port

Comment (fabianvf) [reply example/deploy/operator.yaml]: Ah never mind, I missed the image name change

File: pkg/runner/runner.go (hunk 7 above)
Comment (shawn-hurley) [reply pkg/runner/runner.go]: I think we need a good mechanism for this and should follow up on this.
